Scripts/Runtime/Util/Jobs/ParallelCollectionUtil.cs

```diff
@@ -92,12 +101,37 @@ public static int CollectionSizeForMaxThreads()
         /// <returns>The collection index to use</returns>
         public static int CollectionIndexForThread(int nativeThreadIndex)
         {
-            // JobsUtility.JobWorkerMaximumCount = 15
-            // index  0, 1, 2, 3, 4, 5, 6, 7, 8, 9,  10, 11, 12, 13, 14, 15
-            // thread 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 17
-            return (nativeThreadIndex <= JOB_WORKER_MAXIMUM_COUNT)
-                ? nativeThreadIndex - 1
-                : JOB_WORKER_MAXIMUM_COUNT;
+            DetectMultipleXThreads(nativeThreadIndex);
+            Debug.Assert(nativeThreadIndex > 0 && nativeThreadIndex <= JobsUtility.MaxJobThreadCount);
+            return math.min(nativeThreadIndex - 1, JOB_WORKER_MAXIMUM_COUNT.Data);
+        }
+
+
+        [Conditional("ENABLE_UNITY_COLLECTIONS_CHECKS")]
+        [BurstDiscard]
+        private static void DetectMultipleXThreads(int nativeThreadIndex)
+        {
+            ThreadHelper.DetectMultipleXThreads(nativeThreadIndex, CollectionSizeForMaxThreads);
+        }
+    }
+
+    
+    internal static class ThreadHelper
+    {
+        private static readonly ConcurrentBag<int> s_ThreadIndicesSeen = new ConcurrentBag<int>();
```

Comment:
Does this need to get re-initialized on subsystem registration?

Reply:
Good point, fixed.